File: lin_bin_search/lin_bin_search.c
/*
 * Compare linear and binary search on ordered arrays
 */

#include <stdio.h>
#include <stdlib.h>

int linear_search(int *, size_t, int, int *);
int binary_search(int *, size_t, int, int *);


int main()
{

    int steps = 0;
    int value = 101;
    size_t array_size;
    int array[] = {1, 2, 4, 5, 8, 9, 20, 22, 24, 28, 30, 32, 34, 40, 42, 50, 52,
                   60, 80, 90, 92, 94, 96, 100};
    array_size = sizeof(array) / sizeof(array[0]);

    /* evaluate linear search  */
    int result = linear_search(array, array_size, value, &steps);
    printf("value: %d, steps: %d\n", result, steps);

    /* reset steps to zero to evaluate binary search  */
    steps = 0;

    /* evaluate linear search  */
    result = binary_search(array, array_size, value, &steps);
    printf("value: %d, steps: %d\n", result, steps);

    return 0;
}

int linear_search(int *ordered_array, size_t sz, int search_value, int *steps){

    int i;
    for (i = 0; i < sz; i++) {
        *steps += 1;
        if (search_value == ordered_array[i])
            return i;
        else if (search_value < ordered_array[i])
            break;
    }
    return -1;
}


int binary_search(int *ordered_array, size_t sz, int search_value, int *steps){

    int start = 0;
    int end = sz-1;

    while (start < end){
        *steps += 1;

        int midpoint = (start + end) / 2;

        if (search_value == ordered_array[midpoint])
            return midpoint;
        else if (search_value > ordered_array[midpoint])
            start = midpoint + 1;
        else if (search_value < ordered_array[midpoint])
            end = midpoint - 1;
    }

    return -1;
}

File: python/linear_binary_search/lin_bin_search.py
"""
Compare linear and binary search on ordered arrays
"""

def linear_search(ordered_array, search_value):

    """Return:
       index of search_value, number of steps taken

       If search_value is not in the array return -1 and the number of steps taken

    >>> linear_search([1, 2, 5, 7, 14], 7)
    (3, 4)
    >>> linear_search([1, 2, 5, 7, 14], 4)
    (-1, 3)
    >>> linear_search([1, 2, 5, 7, 14], 15)
    (-1, 5)
    >>> linear_search(range(0, 1000, 2) , 62)
    (31, 32)
    >>> linear_search(range(0, 1000, 2) , 761)
    (-1, 382)
    """
    checks = 0
    for idx, val in enumerate(ordered_array):
        checks += 1
        if val == search_value:
            return idx, checks
        # if we reach an element greater the then search_value
        elif val > search_value:
            break
    return -1, checks


def binary_search(ordered_array, search_value):

    """Return:
       index of search_value, number of steps taken

       If search_value is not in the array return -1 and the number of steps taken

    >>> binary_search([1, 2, 5, 6, 7, 10, 12], 10)
    (5, 2)
    >>> binary_search(range(0, 1000, 2) , 62)
    (31, 8)
    >>> binary_search(range(0, 1000, 2) , 761)
    (-1, 9)
    """
    checks = 0
    start=0
    end=len(ordered_array) - 1

    while start <= end:
        checks += 1
        val_idx = (start + end) // 2
        val_midpoint = ordered_array[val_idx]

        if search_value == val_midpoint:
            return val_idx, checks

        elif search_value < val_midpoint:
            end = val_idx - 1

        elif  search_value > val_midpoint:
            start = val_idx + 1

    return -1, checks

if __name__ == "__main__":
    import doctest
    doctest.testmod()
